File: Exercicios/Ex1_Transporte/Ex1.java
// PRODUTOS ABSTRATOS
abstract class VeiculoIndividual {
    public abstract String iniciarRota();
}

abstract class VeiculoColetivo {
    public abstract String iniciarRota();
}

// PRODUTOS CONCRETOS - TERRESTRE
class Carro extends VeiculoIndividual {
    @Override
    public String iniciarRota() {
        return "Partiu! Carro na pista.";
    }
}

class Onibus extends VeiculoColetivo {
    @Override
    public String iniciarRota() {
        return "Lotou! Ônibus saindo do ponto.";
    }
}

// PRODUTOS CONCRETOS - AEREO
class Helicoptero extends VeiculoIndividual {
    @Override
    public String iniciarRota() {
        return "Hélice girando. Decolagem autorizada!";
    }
}

class Aviao extends VeiculoColetivo {
    @Override
    public String iniciarRota() {
        return "Atenção passageiros, iniciando voo.";
    }
}

// FABRICA ABSTRATA
abstract class FabricaTransporte {
    public abstract VeiculoIndividual criarVeiculoIndividual();

    public abstract VeiculoColetivo criarVeiculoColetivo();
}

// FABRICAS CONCRETAS
class FabricaTerrestre extends FabricaTransporte {
    @Override
    public VeiculoIndividual criarVeiculoIndividual() {
        return new Carro();
    }

    @Override
    public VeiculoColetivo criarVeiculoColetivo() {
        return new Onibus();
    }
}

class FabricaAerea extends FabricaTransporte {
    @Override
    public VeiculoIndividual criarVeiculoIndividual() {
        return new Helicoptero();
    }

    @Override
    public VeiculoColetivo criarVeiculoColetivo() {
        return new Aviao();
    }
}

// CLIENTE
public class Ex1 {
    public static void iniciarTransporte(FabricaTransporte fabrica) {
        VeiculoIndividual individual = fabrica.criarVeiculoIndividual();
        VeiculoColetivo coletivo = fabrica.criarVeiculoColetivo();

        System.out.println(" -> " + individual.iniciarRota());
        System.out.println(" -> " + coletivo.iniciarRota());
        System.out.println(" ----------------------------");
    }

    public static void main(String[] args) {
        System.out.println("\n--- VAI POR TERRA ---");
        iniciarTransporte(new FabricaTerrestre());

        System.out.println("\n--- VAI POR AR ---");
        iniciarTransporte(new FabricaAerea());
    }
}

File: Abstract Factory/Exercicios/Ex1_Transporte/Ex1.py
"""
PADRÃO: ABSTRACT FACTORY (FÁBRICA ABSTRATA)

Exercício: Sistema de Transporte.
Este padrão é ideal quando o sistema deve gerenciar famílias de produtos 
(Individual e Coletivo) sem acoplar o cliente às classes concretas.
"""

from abc import ABC, abstractmethod

# 1. PRODUTOS ABSTRATOS
class VeiculoIndividual(ABC):
    @abstractmethod
    def iniciar_rota(self):
        pass

class VeiculoColetivo(ABC):
    @abstractmethod
    def iniciar_rota(self):
        pass

# 2. PRODUTOS CONCRETOS - FAMÍLIA TERRESTRE
class Carro(VeiculoIndividual):
    def iniciar_rota(self):
        return "Partiu! Carro na pista (Transporte Terrestre Individual)."

class Onibus(VeiculoColetivo):
    def iniciar_rota(self):
        return "Lotou! Ônibus saindo do ponto (Transporte Terrestre Coletivo)."

# 2. PRODUTOS CONCRETOS - FAMÍLIA AÉREA
class Helicoptero(VeiculoIndividual):
    def iniciar_rota(self):
        return "Hélice girando. Decolagem autorizada! (Transporte Aéreo Individual)."

class Aviao(VeiculoColetivo):
    def iniciar_rota(self):
        return "Atenção passageiros, iniciando voo (Transporte Aéreo Coletivo)."

# 3. FABRICA ABSTRATA
class FabricaTransporte(ABC):
    @abstractmethod
    def criar_veiculo_individual(self) -> VeiculoIndividual:
        pass
    
    @abstractmethod
    def criar_veiculo_coletivo(self) -> VeiculoColetivo:
        pass

# 4. FABRICAS CONCRETAS
class FabricaTerrestre(FabricaTransporte):
    def criar_veiculo_individual(self):
        return Carro()
    
    def criar_veiculo_coletivo(self):
        return Onibus()

class FabricaAerea(FabricaTransporte):
    def criar_veiculo_individual(self):
        return Helicoptero()
    
    def criar_veiculo_coletivo(self):
        return Aviao()

# 5. CLIENTE
def iniciar_transporte(fabrica: FabricaTransporte):
    individual = fabrica.criar_veiculo_individual()
    coletivo = fabrica.criar_veiculo_coletivo()

    print(f" -> {individual.iniciar_rota()}")
    print(f" -> {coletivo.iniciar_rota()}")
    print(" ----------------------------")

if __name__ == "__main__":
    print("\n--- AMBIENTE TERRESTRE ---")
    iniciar_transporte(FabricaTerrestre())

    print("\n--- AMBIENTE AÉREO ---")
    iniciar_transporte(FabricaAerea())
